Reduce invasiveness of credit fees call

piersy opened the pull request:
Moving state transition logic into `distributeTxFees` made rebasing much harder for conflicting changes in that area.

This PR seeks to reduce the invasiveness of our call to `CreditFees` such that as much as possible existing code is left unchanged.

I arrived at these changes by looking at the original commit that introduced `distributeTxFees` and then tried to unwind that, see that diff with `git show c0028659016f5d7e838146bc295806a1821365`

One problem that remains is that in the previous implementation the fee handler was credited with the base fee in all contexts, whereas now the fee handler is credited only when the following is true:
```
optimismConfig := st.evm.ChainConfig().Optimism; optimismConfig != nil && rules.IsOptimismBedrock && !st.msg.IsDepositTx 
```

I think this is more correct since we avoid crediting the fee handler for deposit transactions and we should always be in an OptimismBedrock context but this breaks the `test_base_fee_recipient.sh` e2e test since the devchain config does not set optimism config or the bedrock block. So I have skipped `test_base_fee_recipient.sh` for now. I believe the best fix for this would be to make the devchain more closely resemble a real chain by using closer to real config.

Fixes https://github.com/celo-org/celo-blockchain-planning/issues/991

Depends on #381 so must be merged after

piersy committed:
Remove use of distributeTxFees function

distributeTxFees made rebasing hard because it moved and modified large
chunks of state transition logic, making it hard to resolve conflicts
ocurring in the area where code had been moved from.

This change re-instates the old code and adds a block to the if else
statement to handle celo calling CreditFees for fee currency
transactions.

ezdac commented on core/state_transition.go:
```diff
@@ -683,9 +685,101 @@ func (st *stateTransition) innerExecute() (*ExecutionResult, error) {
+			return nil, err
+		}
 	}
+	// if err := st.distributeTxFees(); err != nil {
```
Did you leave this on purpose? Otherwise I'd remove this

ezdac commented on core/state_transition.go:
```diff
@@ -683,9 +685,101 @@ func (st *stateTransition) innerExecute() (*ExecutionResult, error) {
+
+		// Check that we are post bedrock to enable op-geth to be able to create pseudo pre-bedrock blocks (these are pre-bedrock, but don't follow l2 geth rules)
+		// Note optimismConfig will not be nil if rules.IsOptimismBedrock is true
+		if optimismConfig := st.evm.ChainConfig().Optimism; optimismConfig != nil && rules.IsOptimismBedrock && !st.msg.IsDepositTx {
```
You said this is different now - does that mean that deposit transactions resulted in a credit to the fee-handler before? The fee-distribution is skipped earlier already for deposits when the `rules.IsRegolith` is active.
If we ran a chain prior to regolith, introducing this here would be a hard-fork condition (which we didn't?)

piersy replied:
Yes, before all transactions would have been executing the code within (or equivalent code within) this if statement. But now only transactions received post bedrock in an optimism context that are not deposit transactions will cause the code inside the if to be executed.

It looks like all our L2 chains started with Regolith enabled, so as you point out we would be exiting earlier for deposit transactions, so that should not have changed.

What has changed is that this code will no longer execute outside of a non optimism bedrock context, which is something our base fee recipient e2e test relied upon. For our existing chains this shouldn't affect anything since they should always have the optimsim config set and be post bedrock.

ezdac commented on core/state_transition.go:
```diff
@@ -683,9 +685,101 @@ func (st *stateTransition) innerExecute() (*ExecutionResult, error) {
+		gasUsed := new(big.Int).SetUint64(st.gasUsed())
+		totalTxFee := new(big.Int).Mul(gasUsed, st.msg.GasPrice)
+		baseTxFee := new(big.Int).Mul(gasUsed, st.calculateBaseFee())
+		// No need to do effectiveTip calculation, because st.gasPrice == effectiveGasPrice, and effectiveTip = effectiveGasPrice - baseTxFee
```
Why is the calculation necessary above but not here?
Can't we reuse the `effectiveTipU256` from above here?

piersy replied:
We can, it was something carried over from our previous implementation.

ezdac replied:
I left this in, since we would have to calculate the "effectiveTip" deduction again (by multiplying `effectiveTipU256` with the gas used) - what do you think?

ezdac commented on core/state_transition.go:
```diff
@@ -683,9 +685,101 @@ func (st *stateTransition) innerExecute() (*ExecutionResult, error) {
+			if rules.IsCel2 {
+				feeHandlerAddress := addresses.GetAddressesOrDefault(st.evm.ChainConfig().ChainID, addresses.MainnetAddresses).FeeHandler
+				st.state.AddBalance(feeHandlerAddress, amtU256, tracing.BalanceIncreaseRewardTransactionFee)
+			} else if st.evm.ChainConfig().Optimism != nil {
```
this can be an else, the Optimism != nil condition is already checked before

ezdac commented on core/state_transition.go:
```diff
@@ -683,9 +685,101 @@ func (st *stateTransition) innerExecute() (*ExecutionResult, error) {
+			} else if st.evm.ChainConfig().Optimism != nil {
+				st.state.AddBalance(params.OptimismBaseFeeRecipient, amtU256, tracing.BalanceIncreaseRewardTransactionFee)
+			}
+			st.state.AddBalance(params.OptimismBaseFeeRecipient, amtU256, tracing.BalanceIncreaseRewardTransactionFee)
```
This is doubled! **Especially when we set the OptimismBaseFeeRecipient to the fee-handler, it will get the amount twice!** 
This would also be a hard-fork condition in general, since the balances will be off.

Why not leave this in here, but only change the receiving address in the if /else above

ezdac commented on core/state_transition.go:
```diff
@@ -683,9 +685,101 @@ func (st *stateTransition) innerExecute() (*ExecutionResult, error) {
+		// Check that we are post bedrock to enable op-geth to be able to create pseudo pre-bedrock blocks (these are pre-bedrock, but don't follow l2 geth rules)
+		// Note optimismConfig will not be nil if rules.IsOptimismBedrock is true
+		if optimismConfig := st.evm.ChainConfig().Optimism; optimismConfig != nil &&
+			rules.IsOptimismBedrock && !st.msg.IsDepositTx {
```
This means that pre-regolith fee-currency transactions *do* credit the fees for deposit transactions, but non-fee-currency transactions don't. 
I think we need to break out if the whole fee-currency "else" condition if we have a deposit.

ezdac commented on core/state_transition.go:
```diff
@@ -766,12 +860,24 @@ func (st *stateTransition) refundGas(refundQuotient uint64) uint64 {
+	st.state.AddBalance(st.msg.From, remaining, tracing.BalanceIncreaseGasReturn)
+
+	if st.evm.Config.Tracer != nil && st.evm.Config.Tracer.OnGasChange != nil && st.gasRemaining > 0 {
+		st.evm.Config.Tracer.OnGasChange(st.gasRemaining, 0, tracing.GasChangeTxLeftOverReturned)
```
This tracer call will be missing for fee-currency transactions, since the `CreditFees` does not call it

ezdac commented on core/state_transition.go:
```diff
@@ -683,9 +685,101 @@ func (st *stateTransition) innerExecute() (*ExecutionResult, error) {
+			log.Error("Error crediting", "from", from, "coinbase", st.evm.Context.Coinbase, "feeHandler", feeHandlerAddress, "err", err)
+			return nil, err
+		}
 	}
```
You're missing the tracer call here (see comment below in `refundGas` function)

ezdac reviewed (changes requested):
I don't have the best feeling with this PR - in general the function is quite important 
and mistakes here could have more severe consequences.
It took me quite some time to go through everything, so I can see how easy it is to get something wrong here.

The main immediate issue I found is the doubled base-fee distribution for the non-fee-currency.

We should spend some more time to go into this in detail and be extra sure about those changes.

piersy committed:
Fix TestPrestateTracer fee currency tests

We needed to enable optimsim in the config to ensure creditFees was
executed but that had a knock on effect of requireing the L1CostFunc to
also be set, so we now set it.

piersy committed:
Fix TestNativeTransferWithFeeCurrency

piersy committed:
Fix TestNativeTransferWithFeeCurrencyAndTooLowGasPrice

ezdac commented on eth/ethconfig/config.go:
```diff
@@ -190,6 +190,11 @@ type Config struct {
 func CreateConsensusEngine(config *params.ChainConfig, db ethdb.Database) (consensus.Engine, error) {
+	// In dev mode we want to maintain the fake consensus mechanism
+	if config.ChainID != nil && config.ChainID.Int64() == params.DevChainID {
+		return beacon.New(ethash.NewFaker()), nil
```
Removing this does not change the outcome of a local e2e-test and unit-test run for me.

Why did you add this? To me it seems that the passed in `ethone` is never used anyways because all pre-merge conditions are false, so it doesn't seem to matter?

piersy replied:
I added this because in some previous state I thought this was fixing something, but now I can't seem to reproduce any failures by removing it. So lets remove it

ezdac commented on e2e_test/run_all_tests.sh:
```diff
@@ -43,7 +43,18 @@ for f in test_*"$TEST_GLOB"*; do
 		  ;;
-	    esac
+		esac
+	else
```
Since we now have the optimism config, the skip can be removed again.
For me this test passes locally

ezdac commented on params/celo_config.go:
```diff
@@ -72,4 +74,68 @@ var (
+		FjordTime:    newUint64(0),
+		GraniteTime:  newUint64(0),
+		HoloceneTime: nil,
+		IsthmusTime:  nil,
```
We have to remember to set this and PragueTime for the rebased version

piersy replied:
Yeah, I wondered about this, added a task - https://github.com/celo-org/celo-blockchain-planning/issues/1034

ezdac commented on params/celo_config.go:
```diff
@@ -72,4 +74,68 @@ var (
+
+		Celo: &CeloConfig{
+			// 25000000000 is the base fee floor for mainnet, we use a lower value for dev mode
+			// because the real value breaks many our our e2e tests.
```
```suggestion
			// because the real value breaks many of our e2e tests.
```

ezdac reviewed (changes requested):
Nice that so many e2e tests work again with only minimal changes!
I have some minor remarks and didn't understand the need for the dev-chain specific "ethone"  consensus.

piersy commented:
@ezdac I'm going to split changes to provide realistic dev chain config into it's own pr

piersy committed:
Use realistic config for dev mode

The dev mode config now mostly reflects our mainnet config. In order to
support this updated config the SimulatedBeacon was updated to include
gas limit and a system transaction in the fork choice updated payload
attributes.

piersy commented:
Hey @ezdac I've updated this with a force push, it now just contains the unwinding of the creditFees call, running dev mode with realistic config has been moved to #381

piersy committed:
Merge branch 'celo-rebase-12' into piersy/reduce-invasiveness-of-credit-fees-call
